design/design-generic-references.md

```diff
@@ -0,0 +1,314 @@
+A number of useful transformation functions could be made available to that context.
+
+### Advantages
+
+* All referred values are declared on managed resources using already familiar
+  mechanism, keeping their discovery simple
+* Value extraction is detached from its consumption, same `Referable` can be
+  used by many multiple manged resources
+* `Referable` instances can be watched and used by non-Crossplane controllers as
+  well
+* `Referable` instances can be properly garbage collected if owner
+  or [`Usage`](https://github.com/crossplane/crossplane/blob/master/design/one-pager-generic-usage-type.md)
+  instances are declared
+* Multiple objects/fields can be referred as value sources
+* Value can be transformed using a Go template
+
+### Disadvantages
+
+* Providers might need to get read access to `Referable` type
+
+### Implementation
+
+It consists of two parts:
+
+* Adding new Crossplane controller for `Referable` type
+* Update [controller-runtime](https://github.com/crossplane/crossplane-runtime/)
+  to support usage of `Referable` instances
+
+#### Referable Controller
+
+The controller is responsible for:
+
+* Defining proper `Usage` instance to guard against improper deletion
+  of `Referable` instances
+* Retrieving the declared objects and the field value, based on the provided
+  references
+* Transforming the value (if requested) and exposing it under `status.value`
+* Marking the instance as ready
+
+We assume currently that the value does not change after it gets exposed.
+
+No additional RBAC rules are needed, since Crossplane already poses very broad
+permissions within the cluster.
+
+#### Controller-runtime Changes
+
+* [APIResolver](https://github.com/crossplane/crossplane-runtime/blob/master/pkg/reference/reference.go#L280)
+  should be enriched to support `Referable` instances
+* If additional references need to be declared via `spec.refs` block or
+  annotations, [ResolveReferences function generator](https://github.com/crossplane/crossplane-tools/blob/master/internal/method/resolver.go#L33)
+  should be enriched to support them
+* After fetching values, proper `Usage` instance need to be defined to guard
+  against improper deletion
+
+Finally, providers need to be upgraded to use the new version of controller-runtime 
+and each type `ResolveReference` function needs to be regenerated.
+
+## Alternatives Considered
+
+* There is [a previous work on this topic](https://github.com/crossplane/crossplane/pull/2385)  
+  It embeds the schema similar `Referable` type into a number of `spec.patches.fromObject`
+  fields in managed resource. Such approach would require that each provider 
+  has access to broad set of resources (even from other providers), demanding
+  a broad set of RBAC rules. Furthermore, if a certain value transformation is
+  needed on multiple managed resources, it would be required to repeat its 
+  declaration on every MR.
```

Comment:
```suggestion
* [A previous design](https://github.com/crossplane/crossplane/pull/2385) was proposed for this topic.
  It embeds the schema similar `Referable` type into a number of `spec.patches.fromObject`
  fields in managed resources. Such an approach would require that each provider 
  has access to a broad set of resources (even from other providers), demanding
  a broad set of RBAC rules. Furthermore, if a certain value transformation is
  needed on multiple managed resources, it would be required to repeat its 
  declaration on every MR.
```